### DISCUSSION TITLE
Making a local error handler using `@error`

### DISCUSSION THREAD

**Discussion by hitblast** [Rapptz/discord.py#8404] [Q&A] [answered]
While coding a cog for handling exceptions inside my own bot, I stumbled upon some new stuff inside the documentation. If somebody could please explain the use of [`@error`](https://discordpy.readthedocs.io/en/stable/interactions/api.html?highlight=commandtree%20error#discord.app_commands.CommandTree.error) and how it can be used inside a cog to develop a rather *separated* exception handler then it'd be pretty helpful. Cheers :D

**Comment by Soheab** [comment, marked as answer]
The `@CommandTree.error` decorator cannot be used inside a cog but there are alternatives.

1. [`Cog.cog_app_command_error`](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html#discord.ext.commands.Cog.cog_app_command_error), this is a special method that is called for errors that occur for app commands that are defined in that cog.
```py
from discord import app_commands, Interaction 
from discord.ext import commands

class MyCog(commands.Cog):
    async def cog_app_command_error(
        self,
        interaction: Interaction,
        error: app_commands.AppCommandError
    ):
        # disclaimer: this is an example implementation.
        print("An error occurred in the following command:", interaction.command, "error:", str(error))
        raise error
```
2. Staying on the theme of in a Cog but not really recommended, you can override the [`CommandTree.on_error`](https://discordpy.readthedocs.io/en/stable/interactions/api.html#discord.app_commands.CommandTree.on_error) method by reassigning it, this will function as a global error handler for all app commands.
```py
from discord import app_commands, Interaction 
from discord.ext import commands

class MyCog(commands.Cog):
    def __init__(self, bot: commands.Bot):
        ...
        # assign the handler
        bot.tree.on_error = self.global_app_command_error
        
    async def global_app_command_error(
        self,
        interaction: Interaction,
        error: app_commands.AppCommandError
    ):
        # disclaimer: this is an example implementation.
        print("An error occurred in the following command:", interaction.command, "error:", str(error))
        raise error

async def setup(bot: commands.Bot):
    await bot.add_cog(MyCog(bot))
```
3. Subclass [`CommandTree`](https://discordpy.readthedocs.io/en/stable/interactions/api.html#discord.app_commands.CommandTree) to override `on_error`, this will function as a global error handler for all app commands. **This is** the **recommended** solution.
```py
from discord import app_commands, Interaction 

# subclassing CommandTree
class MyCommandTree(app_commands.CommandTree):
    # overriding on_error
    async def on_error(
        self,
        interaction: Interaction,
        error: app_commands.AppCommandError
    ):
        # disclaimer: this is an example implementation.
        print("An error occurred in the following command:", interaction.command, "error:", str(error))
        raise error

# using it
# discord.Client
# tree = MyCommandTree(client)

# -> commands.Bot
bot = commands.Bot(
    command_prefix=...,
    intents=...,
    tree_cls=MyCommandTree # passing the class, not calling it.
)
```

**Reply by hitblast** [reply]
Thanks! :D
